mods/mc_worldmanager/realm/realm.lua

```diff
@@ -51,67 +51,156 @@ function Realm:New(name, size, height)
+    local gridStartPos, gridEndPos = Realm.CalculateStartEndPosition(area)
 
     -- Calculate our world position based on our location on the realm grid
-    this.StartPos.x = -20000 + (realmSize * realmLocation.x) + (realmBuffer * realmLocation.x)
-    this.StartPos.z = -20000 + (realmSize * realmLocation.z) + (realmBuffer * realmLocation.z)
+    this.StartPos = Realm.gridToWorldSpace(gridStartPos)
+    this.EndPos = Realm.gridToWorldSpace(gridEndPos)
 
-    -- Ensures that a realm size is no larger than our maximum size.
-    local finalRealmSize = math.min(realmSize, size)
-    local finalRealmHeight = math.min(realmHeight, height)
-
-    this.EndPos = { x = this.StartPos.x + finalRealmSize,
-                    y = this.StartPos.y + finalRealmHeight,
-                    z = this.StartPos.z + finalRealmSize }
 
     -- Temporary spawn point calculation
     this.SpawnPoint = { x = (this.StartPos.x + this.EndPos.x) / 2,
-                        y = ((this.StartPos.y + this.EndPos.y) / 2) + 2,
+                        y = (this.StartPos.y + 2),
                         z = (this.StartPos.z + this.EndPos.z) / 2 }
 
     setmetatable(this, self)
-    Realm.realmDict[self.ID] = this
+    Realm.realmDict[this.ID] = this
+
     Realm.SaveDataToStorage()
 
     return this
 end
 
+-- Online bin packing... A pretty challenging problem to solve.
+-- To simplify the problem, we'll create new bins for each new realm we make.
+-- When we need to create new realms, we'll run the bin packing algorithm
+-- If the realm doesn't fit into any of the existing bins, we'll create a new one.
+
+-- We'll use a mapchunk has the smallest denominator for world delineation
+-- What we'll do is keep track of the last realm position as well as areas where realms were deleted.
+-- When creating a new realm, we'll see if the empty space we have can contain a new realm. If it can't,
+-- We'll create the realm after the last realm position.
+Realm.lastRealmPosition = { xStart = Realm.const.bufferSize,
+                            yStart = Realm.const.bufferSize,
+                            zStart = Realm.const.bufferSize }
+
+Realm.maxRealmSize = { x = 0, y = 0, z = 0 }
+Realm.EmptyChunks = {}
+
+function Realm.CalculateStartEndPosition(areaInBlocks)
+
+    -- Note that all of the coordinates used in this function are in "gridSpace"
+    -- This roughly correlates to the chunk coordinates in MineTest
+    -- 1 unit in gridSpace is 80 blocks in worldSpace
+
+    -- calculate our realm size in grid units
+    local realmSize = { x = math.ceil(areaInBlocks.x / 80),
+                        y = math.ceil(areaInBlocks.y / 80),
+                        z = math.ceil(areaInBlocks.z / 80) }
+
+    local reuseBin = false
+    local StartPos = { x = 0, y = 0, z = 0 }
+    local BinEndPos = { x = 0, y = 0, z = 0 }
+
+    for i, v in ipairs(Realm.EmptyChunks) do
+        if (v.area.x >= realmSize.x and v.area.y >= realmSize.y and v.area.z >= realmSize.z) then
```

Comment:
Does this include the buffer distance needed between realms? Not sure if `realmSize.x` is the absolute size of the schematic or if it also includes the buffer or if the buffer distance is checked/included elsewhere.

Reply:
Good catch; I was working on this part right before merging with develop. The buffer zone was not initially included in the empty chunks -- meaning that placement was already determined.  I've since changed this, but have forgotten to update it here.